fs-storage/src/file_storage.rs

```diff
@@ -0,0 +1,246 @@
+use std::fmt::Debug;
+use std::fs::{self, File};
+use std::io::{BufRead, BufReader, BufWriter, Write};
+use std::str::FromStr;
+use std::time::SystemTime;
+use std::{
+    collections::HashMap,
+    fmt::Display,
+    path::{Path, PathBuf},
+};
+
+use data_error::{ArklibError, Result};
+
+const LOG_PREFIX: &str = "[file-storage]";
+const STORAGE_VERSION: i32 = 2;
+const STORAGE_VERSION_PREFIX: &str = "version ";
+const KEY_VALUE_SEPARATOR: char = ':';
+
+pub struct FileStorage {
+    log_prefix: String,
+    label: String,
+    path: PathBuf,
+    timestamp: SystemTime,
+}
```

Comment:
What about such API?
```
pub struct Mapping<K,V> {
    timestamp: SystemTime,
    values: HashMap<K, V>
}

pub struct Storage<K,V> {
    label: String,
    path: PathBuf,
    mapping: Option<Mapping<K,V>> // `None` means not loaded yet
}

impl Storage {
...
    fn new(label: String, path: &Path) -> Self { ... }
    fn load(self: Self) -> Self { ... }
    fn update(self: Self) -> Self { ... }
...
}
```

Or something like this:
```
pub struct Mapping<K,V> {
    timestamp: SystemTime,
    values: HashMap<K, V>
}

pub struct Storage<K,V> {
    label: String,
    path: PathBuf,
    mapping: Mapping<K,V>
}

impl Storage {
...
    fn provide(label: String, path: &Path) -> Self { ... }
    fn update(self: Self) -> Self { ... }
...
}
```

Reply:
I've updated the api to remove the callback style and just return the hashmap directly. Currently I don't see how the additional complexity of wrapping the HashMap in a Mapping will help.

I think the current state of the PR is a good first version of a functional file storage implementation. I propose we merge it and add more features as separate issues. I prefer not to keep PRs open for too long because they lead to merge conflicts.

Reply:
Agree about the conflicts. But if the changes are small it's better to make it in same PR to avoid excessive task management.

`Mapping` wrapper is useful only because it also contains timestamp. We could use a tuple, but named structure is better. The rest looks good for now.